File: plugin-tester-scala/src/main/scala/example/myapp/helloworld/RestartingGreeterClient.scala
package example.myapp.helloworld

import akka.{Done, NotUsed}
import akka.actor.ActorSystem
import akka.grpc.GrpcClientSettings
import akka.grpc.scaladsl.RestartingClient
import akka.stream.ActorMaterializer
import akka.stream.scaladsl.Source
import example.myapp.helloworld.grpc._

import scala.concurrent.{Await, Future}
import scala.concurrent.duration._
import scala.util.Try

object RestartingGreeterClient {

  def main(args: Array[String]): Unit = {

    implicit val sys = ActorSystem()
    implicit val mat = ActorMaterializer()
    implicit val ec = sys.dispatcher

    //#restarting-client
    // Function for creating a client
    val clientSettings = GrpcClientSettings.fromConfig(GreeterService.name)
    val clientConstructor = () => new GreeterServiceClient(clientSettings)

    // Wrapped in a restarting client
    val restartingClient = new RestartingClient[GreeterServiceClient](clientConstructor)
    //#restarting-client

    singleRequestReply()
    streamingRequest()
    streamingReply()
    streamingRequestReply()

    sys.scheduler.schedule(1.second, 1.second) {
      Try(singleRequestReply())
    }

    def singleRequestReply(): Unit = {
      sys.log.info("Performing request")

      //#usage
      val reply = restartingClient.withClient(c => c.sayHello(HelloRequest("Bob")))
      //#usage

      println(s"got single reply: ${Await.result(reply, 5.seconds).message}")
    }

    def streamingRequest(): Unit = {
      val requests = List("Alice", "Bob", "Peter").map(HelloRequest.apply)
      val reply = restartingClient.withClient(c => c.itKeepsTalking(Source(requests)))
      println(s"got single reply for streaming requests: ${Await.result(reply, 5.seconds).message}")
    }

    def streamingReply(): Unit = {
      val responseStream = restartingClient.withClient(c => c.itKeepsReplying(HelloRequest("Alice")))
      val done: Future[Done] =
        responseStream.runForeach(reply => println(s"got streaming reply: ${reply.message}"))
      Await.ready(done, 1.minute) // just to keep sample simple - don't do Await.ready in actual code
    }

    def streamingRequestReply(): Unit = {
      val requestStream: Source[HelloRequest, NotUsed] =
        Source
          .tick(100.millis, 1.second, "tick")
          .zipWithIndex
          .map { case (_, i) => i }
          .map(i => HelloRequest(s"Alice-$i"))
          .take(10)
          .mapMaterializedValue(_ => NotUsed)

      val responseStream: Source[HelloReply, NotUsed] = restartingClient.withClient(c => c.streamHellos(requestStream))
      val done: Future[Done] =
        responseStream.runForeach(reply => println(s"got streaming reply: ${reply.message}"))
      Await.ready(done, 1.minute)
    }
  }

}

File: plugin-tester-java/src/main/java/example/myapp/helloworld/RestartingGreeterClient.java
package example.myapp.helloworld;

import akka.Done;
import akka.NotUsed;
import akka.actor.ActorSystem;
import akka.grpc.GrpcClientSettings;
import akka.grpc.javadsl.RestartingClient;
import akka.japi.Pair;
import akka.stream.ActorMaterializer;
import akka.stream.Materializer;
import akka.stream.javadsl.Source;

import java.time.Duration;
import java.util.List;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import example.myapp.helloworld.grpc.*;

class RestartingGreeterClient {
    public static void main(String[] args) throws Exception {

        String serverHost = "127.0.0.1";
        int serverPort = 8080;

        ActorSystem system = ActorSystem.create();
        Materializer materializer = ActorMaterializer.create(system);

        try {
            //#restarting-client
            GrpcClientSettings settings = GrpcClientSettings.fromConfig(GreeterService.name, system);

            RestartingClient<GreeterServiceClient> client = new RestartingClient<>(
                    () -> GreeterServiceClient.create(settings, materializer, system.dispatcher()), system.dispatcher()
            );
            //#restarting-client

            singleRequestReply(client);
            streamingRequest(client);
            streamingReply(client, materializer);
            streamingRequestReply(client, materializer);

            client.close();


        } finally {
            system.terminate();
        }

    }

    private static void singleRequestReply(RestartingClient<GreeterServiceClient> client) throws Exception {
        //#usage
        HelloRequest request = HelloRequest.newBuilder().setName("Alice").build();
        CompletionStage<HelloReply> reply = client.withClient(c -> c.sayHello(request));
        System.out.println("got single reply: " + reply.toCompletableFuture().get(5, TimeUnit.SECONDS));
        //#usage
    }

    private static void streamingRequest(RestartingClient<GreeterServiceClient> client) throws Exception {
        List<HelloRequest> requests = Stream.of("Alice", "Bob", "Peter")
                .map(name -> HelloRequest.newBuilder().setName(name).build())
                .collect(Collectors.toList());
        CompletionStage<HelloReply> reply = client.withClient(c -> c.itKeepsTalking(Source.from(requests)));
        System.out.println("got single reply for streaming requests: " +
                reply.toCompletableFuture().get(5, TimeUnit.SECONDS));
    }

    private static void streamingReply(RestartingClient<GreeterServiceClient> client, Materializer mat) throws Exception {
        HelloRequest request = HelloRequest.newBuilder().setName("Alice").build();
        Source<HelloReply, NotUsed> responseStream = client.withClient(c -> c.itKeepsReplying(request));
        CompletionStage<Done> done =
                responseStream.runForeach(reply ->
                        System.out.println("got streaming reply: " + reply.getMessage()), mat);

        done.toCompletableFuture().get(60, TimeUnit.SECONDS);
    }

    private static void streamingRequestReply(RestartingClient<GreeterServiceClient> client, Materializer mat) throws Exception {
        Duration interval = Duration.ofSeconds(1);
        Source<HelloRequest, NotUsed> requestStream = Source
                .tick(interval, interval, "tick")
                .zipWithIndex()
                .map(Pair::second)
                .map(i -> HelloRequest.newBuilder().setName("Alice-" + i).build())
                .take(10)
                .mapMaterializedValue(m -> NotUsed.getInstance());

        Source<HelloReply, NotUsed> responseStream = client.withClient(c -> c.streamHellos(requestStream));
        CompletionStage<Done> done =
                responseStream.runForeach(reply ->
                        System.out.println("got streaming reply: " + reply.getMessage()), mat);

        done.toCompletableFuture().get(60, TimeUnit.SECONDS);
    }

}
